File: FinanceApp/FinanceApp/ClientApp/src/app/models/Loan.ts
import {CurrencyType} from "./Currency";

export class Loan {
  constructor(id: string, name: string, otherParty: string, amount: number, currency: CurrencyType, termInMonths: number, interestRate: number) {
    this.id = id;
    this.name = name;
    this.otherParty = otherParty;
    this.amount = amount;
    this.currency = currency;
    this.interestRate = interestRate;
    this.termInMonths = termInMonths;
  }

  id: string;
  name: string;
  otherParty: string;
  amount: number;
  currency: CurrencyType;
  termInMonths: number;
  interestRate: number;
  monthlyInstallment: number = 0;
  remainingMonths: number = 0;
  remainingAmount: number = 0;
  loanSchedule: LoanSchedule[] = [];

}

export class LoanSchedule {
  index: number = 0;
  startingBalance: number = 0;
  interest: number = 0;
  principal: number = 0;
  endingBalance: number = 0;
}

File: FinanceApp/FinanceApp/Models/Loan.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;
using FinanceApp.Enums;
using FinanceApp.Other;

namespace FinanceApp.Models;

public class Loan
{
    [Required] [Key] public int Id { get; set; }

    [Required] public string Name { get; set; }

    [Required] public string OtherParty { get; set; }

    [Required] public double Amount { get; set; }

    [Required] public Currency Currency { get; set; }

    [Required] public int TermInMonths { get; set; }

    [Required] public double InterestRate { get; set; }

    public int RemainingMonths { get; set; }

    public double RemainingAmount { get; set; }

    [JsonInclude]
    [NotMapped]
    public double MonthlyInstallment =>
        Math.Round(HelperMethods.CalculateLoanAmortizationAmount(RemainingAmount, InterestRate / 100, RemainingMonths));

    [JsonInclude]
    [NotMapped]
    public List<LoanAmortizationScheduleEntry> LoanSchedule =>
        HelperMethods.CalculateLoanSchedule(RemainingAmount, InterestRate / 100, RemainingMonths, 10);

    [JsonIgnore] public ApplicationUser? User { get; set; }
    [JsonIgnore] [ForeignKey("User")] public string? UserId { get; set; }
}